### PR TITLE
D3ASIM-3464: Initial implementation of saving KPI.

### PR DESCRIPTION
**Pull request by faizan2590**

### CODE REVIEW

File: src/d3a/d3a_core/sim_results/endpoint_buffer.py
Comment (Hassan754) [inline src/d3a/d3a_core/sim_results/endpoint_buffer.py:151]:
```suggestion
        if hasattr(area.current_market, 'const_fee_rate') is not None and \
                area.current_market.const_fee_rate is not None:
```

File: src/d3a/d3a_core/sim_results/endpoint_buffer.py
Comment (spyrostz) [inline src/d3a/d3a_core/sim_results/endpoint_buffer.py:156]: I think that it is better to fetch the real current market maker rate from the strategy instead of the constant, reason for that is because there are plans in the future to add market maker profiles, and multiple market makers per simulation.

**Review by spyrostz — COMMENTED** [review]
One important comment, LGTM once resolved

File: setup.py
Comment (fievelk) [inline setup.py:5]: Reminder to restore the original version

File: src/d3a/d3a_core/sim_results/endpoint_buffer.py
Comment (fievelk) [inline src/d3a/d3a_core/sim_results/endpoint_buffer.py:150]: Please use `"` over `'`. The same comment applies to the rest of the PR.

File: src/d3a/d3a_core/sim_results/endpoint_buffer.py
Comment (fievelk) [inline src/d3a/d3a_core/sim_results/endpoint_buffer.py:152]: Nit: please use `( )` over `\` for line continuation.

File: src/d3a/d3a_core/sim_results/endpoint_buffer.py
Comment (faizan2590) [reply src/d3a/d3a_core/sim_results/endpoint_buffer.py:151]: Can you please explain the motive behind the proposed change?

File: src/d3a/d3a_core/sim_results/endpoint_buffer.py
Comment (faizan2590) [reply src/d3a/d3a_core/sim_results/endpoint_buffer.py:156]: How about using `GlobalConfig.market_maker_rate` ?

File: src/d3a/d3a_core/sim_results/endpoint_buffer.py
Comment (Hassan754) [reply src/d3a/d3a_core/sim_results/endpoint_buffer.py:151]: In your code you are checking if `hasattr(area.current_market, 'const_fee_rate') is not None`, this is not a valid check, the `hasattr` will return a boolean either True or False, and there is no way it will return a None (The value you're comparing against)
Furthermore, the same check is intended to validate if area.current_market has an attrribute `const_fee_rate`, so the second condition `area.current_market is not None` doesn't make sense at this stage, if there is a possibility that this can be None it should be the first condition

Comment (faizan2590) [reply src/d3a/d3a_core/sim_results/endpoint_buffer.py:151]:
```suggestion
        if (getattr(area.current_market, "const_fee_rate", None) is not None and
                area.current_market is not None):
```
How about this? 

`area.current_market is not None` is needed when its called at the start of simulation and `current_market` is None at that point.

File: src/d3a/d3a_core/sim_results/endpoint_buffer.py
Comment (spyrostz) [reply src/d3a/d3a_core/sim_results/endpoint_buffer.py:156]: This will work for now, however keep in mind that this is a dict, therefore you have to access the current market slot value of the market maker rate and not just add the whole dict.

File: integration_tests/savings_kpi.feature
Comment (fievelk) [inline integration_tests/savings_kpi.feature:19]: Why not using behave's [scenario outlines](https://behave.readthedocs.io/en/stable/tutorial.html#scenario-outlines) here?
We already used them in other tests and they're useful to reduce code repetition.

File: integration_tests/steps/kpi.py
Comment (fievelk) [inline integration_tests/steps/kpi.py:39]: Leftover from debug

File: setup.py
Comment (fievelk) [reply setup.py:5]: Another reminder, since you've asked for a new review

File: src/d3a/constants.py
Comment (fievelk) [inline src/d3a/constants.py:50]: Nit: why two blank lines?
```suggestion

FEED_IN_TARIFF = 20
```

File: src/d3a/d3a_core/sim_results/endpoint_buffer.py
Comment (fievelk) [inline src/d3a/d3a_core/sim_results/endpoint_buffer.py:152]: The order of the two operands should be switched:
```suggestion
        if area.current_market and getattr(area.current_market, "const_fee_rate", None) is not None:
```
Besides this: could you explain why we are using `getattr` and `hasattr` to check if the area has these attributes? What's the situation in which we don't have them?

File: src/d3a/d3a_core/sim_results/endpoint_buffer.py
Comment (fievelk) [inline src/d3a/d3a_core/sim_results/endpoint_buffer.py:154]: You've already checked whether `area.current_market.const_fee_rate` is not `None` on the line above this. So I think this could be:
```suggestion
            core_stats_dict["const_fee_rate"] = area.current_market.const_fee_rate
```
Do you agree?

File: src/d3a/d3a_core/sim_results/endpoint_buffer.py
Comment (fievelk) [inline src/d3a/d3a_core/sim_results/endpoint_buffer.py:155]: You've already checked whether `area.current_market.const_fee_rate` is not `None` on the line above this. So I think this could be:
```suggestion
            core_stats_dict["const_fee_rate"] = area.current_market.const_fee_rate
```
Do you agree?

File: src/d3a/d3a_core/sim_results/endpoint_buffer.py
Comment (fievelk) [inline src/d3a/d3a_core/sim_results/endpoint_buffer.py:160]: Nit: please use `(` and `)` to replace `\` for line continuation.

File: src/d3a/setup/kpi/d3asim_3464_family_with_pv.py
Comment (fievelk) [inline src/d3a/setup/kpi/d3asim_3464_family_with_pv.py:16]: Nit: please use `"` over `'`.

Comment (fievelk) [reply src/d3a/setup/kpi/d3asim_3464_family_with_pv.py:16]: (This applies to the entire module).

**Comment by fievelk** [comment]
Left some comments. Will review again once answered.

File: src/d3a/d3a_core/sim_results/endpoint_buffer.py
Comment (hannesdiedrich) [inline src/d3a/d3a_core/sim_results/endpoint_buffer.py:161]: There is a function that does exactly that: https://github.com/gridsingularity/d3a/blob/3c75998cb2b9963030bfebbea5ef4cda6cc625c3/src/d3a/d3a_core/util.py#L502
(BTW; you could refactor the input value `next_market` into `market`, thanks!)

File: src/d3a/constants.py
Comment (hannesdiedrich) [inline src/d3a/constants.py:49]: We want to change this only in a blue moon or why is this not a GlobalConfig parameter? Not an expert here, but doesn't this price also change frequently?

**Review by hannesdiedrich — APPROVED** [review]
One global comment, but probably reasoned by design of this feature. 
LGTM

File: src/d3a/constants.py
Comment (faizan2590) [reply src/d3a/constants.py:49]: > We want to change this only in a blue moon or why is this not a GlobalConfig parameter?

Looks better there, will move it even now

> Not an expert here, but doesn't this price also change frequently?

I haven't changed it yet as for all three referenced cases, feed-in tariff had the same rate.

**Review by hannesdiedrich — APPROVED** [review]
LGTM

**Comment by hannesdiedrich** [comment]
@faizan2590 Just a general question: Is this KPI available to the user (UI) even now without a change in the d3a-web? The ticket suggests that they want to add a bar chart for that.

**Comment by faizan2590** [comment]
> @faizan2590 Just a general question: Is this KPI available to the user (UI) even now without a change in the d3a-web? The ticket suggests that they want to add a bar chart for that.

Yes, give this commit a look: https://github.com/gridsingularity/d3a-interface/pull/223/commits/9371f0858d93922c2cb5e6bb24975c08baa4a080

**Review by spyrostz — DISMISSED** [review]
LGTM, as Pierpaolo also says don't forget to restore the d3a-interface branch to master before merging !

**Pull request merged**
